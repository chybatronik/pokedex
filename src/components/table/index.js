import React, { Component } from 'react'
import { Table, Row, Alert, Badge } from 'reactstrap'
import InfiniteScroll from 'react-infinite-scroller'
import Spinner from 'react-spinkit'

export default class TableComponent extends Component {
  renderRow (item) {
    return (
      <tr key={item.name}>
        <th width='10%' scope='row'> { item.index } </th>
        {
          item.isFetch ? (
            <td width='10%'><Spinner name='circle' color='blue' /></td>
          ) : <td style={{padding:0}} width='10%'><img async src={item.sprites.front_default} height={52} /></td>

        }
        <td>{item.name}</td>
        <td>{item.types && item.types.map((type) => {return (<Badge color='light'>{type.type.name}</Badge>)})}</td>
        <td width='10%'>{item.weight}</td>
        <td width='10%'>{item.height}</td>

      </tr>
    )
  }

  renderRows (array) {
    if (array && array.length) {
      return array.map((item, index) => {
        item.index = index + 1
        if (item.name) {
          return this.renderRow(item)
        }
      })
    } else {
      return []
    }
  }

  renderError () {
    return (
      <Alert color='danger' style={{width: '100%'}}>
        { 'Ошибка подключения к серверу pokeapi.co' }
      </Alert>
    )
  }

  renderLoader () {
    return (
      <Row className='d-flex justify-content-center' key={'loader'}>
        <Spinner name='ball-scale-ripple' color='fuchsia' />
      </Row>
    )
  }

  render () {
    if (this.props.isError) {
      return this.renderError()
    }

    return (
      <InfiniteScroll
        element={'article'}
        pageStart={0}
        loadMore={this.props.loadFunc}
        hasMore={this.props.hasMore}
        loader={this.renderLoader()}
      >
        <Table striped>
          <thead>
            <tr>
              <th>#</th>
              <th>@</th>
              <th>name</th>
              <th>types</th>
              <th>weight</th>
              <th>height</th>
            </tr>
          </thead>
          <tbody>
            { this.renderRows(this.props.array) }
          </tbody>
        </Table>
      </InfiniteScroll>
    )
  }
}
