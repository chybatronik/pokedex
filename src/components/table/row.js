// @flow
import React, { PureComponent } from 'react'
import Spinner from 'react-spinkit'
import { Badge } from 'reactstrap'

type Props = {
  name: string,
  index: number,
  sprites: ?Object,
  types: ?Array<Object>,
  weight: ?number,
  height: ?number,
  isFetch: bool
}

export default class RowComponent extends PureComponent<Props> {
  render () {
    const { name, index, sprites, types, weight, height, isFetch } = this.props
    return (
      <tr key={name}>
        <th width='10%' scope='row' > { index } </th>
        {
          isFetch ? (
            <td width='10%'><Spinner name='circle' color='blue' /></td>
          ) :
          (
            <td style={{padding: 0}} width='10%'>
              <img async src={sprites ? sprites.front_default : null} alt={name} height={52} />
            </td>
          )
        }
        <td>{name}</td>
        <td><TagsComponent types={types}/></td>
        <td width='10%'>{weight}</td>
        <td width='10%'>{height}</td>
      </tr>
    )
  }
}

type PropsTags = {
  types: ?Array<Object>
}

class TagsComponent extends PureComponent<PropsTags> {
  render () {
    const { types } = this.props
    let render_array = []
    if( types && types.length && types.length > 0 ) {
      render_array = types.map((type) => {
        return (<Badge color='light' key={type.type.name}>{type.type.name}</Badge>)
      })
    }
    return render_array
  }
}
