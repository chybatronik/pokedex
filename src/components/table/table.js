// @flow
import React, { PureComponent } from 'react'
import { Table, Row, Alert } from 'reactstrap'
import InfiniteScroll from 'react-infinite-scroller'
import Spinner from 'react-spinkit'
import TableRow from './row'

type Props = {
  isError: bool,
  loadFunc: Function,
  hasMore: bool,
  array: Array<Object>
}
export default class TableComponent extends PureComponent<Props> {
  renderRow (item: Object) {
    return (
      <TableRow {...item} key={'row_' + item.name} />
    )
  }

  renderRows (array: Array<Object>) {
    // if (array && array.length) {
    return array.map((item, index) => {
      item.index = index + 1
      return this.renderRow(item)
      // if (item.name) {
      // }
    })
    // }
  }

  renderError () {
    return (
      <Alert color='danger' style={{width: '100%'}}>
        { 'Ошибка подключения к серверу pokeapi.co' }
      </Alert>
    )
  }

  renderLoader () {
    return (
      <Row className='d-flex justify-content-center' key={'loader'}>
        <Spinner name='ball-scale-ripple' color='fuchsia' />
      </Row>
    )
  }

  render () {
    if (this.props.isError) {
      return this.renderError()
    }

    return (
      <InfiniteScroll
        element={'article'}
        pageStart={0}
        loadMore={this.props.loadFunc}
        hasMore={this.props.hasMore}
        loader={this.renderLoader()}
      >
        <Table striped>
          <thead>
            <tr>
              <th>#</th>
              <th>@</th>
              <th>name</th>
              <th>types</th>
              <th>weight</th>
              <th>height</th>
            </tr>
          </thead>
          <tbody>
            { this.renderRows(this.props.array) }
          </tbody>
        </Table>
      </InfiniteScroll>
    )
  }
}
