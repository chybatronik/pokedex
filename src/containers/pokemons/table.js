// @flow
import React, { Component } from 'react'
import { connect } from 'react-redux'
import Table from '../../components/table'
import { getPokemons } from '../../store/actions/pokemons'
import { selectIndexPokemons } from '../../store/selectors/pokemons'

type Props = {
  list_pokemons: Array<Object>,
  config: Object,
  getData: Function
}

export class TableContainer extends Component<Props> {
  loadFunc = () => {
    const { getData, config } = this.props
    if (!config.isFetch && !config.isError) {
      getData(
        config.offset + config.limit,
        config.limit
      )
    }
  }
  render () {
    const { list_pokemons, config } = this.props
    return (
      <Table
        array={list_pokemons}
        loadFunc={this.loadFunc}
        hasMore={config.hasMore}
        isError={config.isError}
        error={config.error}
      />
    )
  }
}
const mapDispatchToProps = ( dispatch: Function ) => ({
  getData: (offset, limit) => dispatch(getPokemons(offset, limit))
})

const mapStateToProps = state => {
  return {
    // pokemons: {
    //   index: selectIndexPokemons(state),
    //   config: state.pokemons.config
    // },
    list_pokemons: selectIndexPokemons(state),
    config: state.pokemons.config
    // fullPokemons: state.fullPokemons,
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TableContainer)
