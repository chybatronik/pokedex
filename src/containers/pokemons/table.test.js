import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import { shallowToJson } from 'enzyme-to-json'
import { TableContainer } from './table'
import expect from 'expect'

Enzyme.configure({ adapter: new Adapter() })

it('renders table container', () => {
  const pokemons = { index: [], config: {hasMore: true, isError: false, isFetch: false} }
  const mockCallback = jest.fn()
  const output = shallow(
    <TableContainer
      list_pokemons={pokemons.index}
      config={pokemons.config}
      getData= {mockCallback}
    />
  )
  expect(shallowToJson(output)).toMatchSnapshot()
})

it('table change config is isFetch', () => {
  const pokemons = { index: [], config: {hasMore: true, isError: true, isFetch: false} }
  const mockCallback = jest.fn()
  const output = shallow(
    <TableContainer
      list_pokemons={pokemons.index}
      config={pokemons.config}
      getData= {mockCallback}
    />
  )
  expect(shallowToJson(output)).toMatchSnapshot()
})

it('table change config is isError', () => {
  const pokemons = { index: [], config: {hasMore: true, isError: false, isFetch: true} }
  const mockCallback = jest.fn()
  const output = shallow(
    <TableContainer
      list_pokemons={pokemons.index}
      config={pokemons.config}
      getData= {mockCallback}
    />
  )
  expect(shallowToJson(output)).toMatchSnapshot()
})
