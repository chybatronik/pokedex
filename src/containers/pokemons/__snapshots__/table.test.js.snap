// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders table container 1`] = `
<TableComponent
  array={Array []}
  hasMore={true}
  isError={false}
  loadFunc={[Function]}
/>
`;

exports[`table change config is isError 1`] = `
<TableComponent
  array={Array []}
  hasMore={true}
  isError={false}
  loadFunc={[Function]}
/>
`;

exports[`table change config is isFetch 1`] = `
<TableComponent
  array={Array []}
  hasMore={true}
  isError={true}
  loadFunc={[Function]}
/>
`;
